core/types/permissions_cache.go

```diff
@@ -0,0 +1,406 @@
+}
+
+func (a *AcctCache) GetAcctList() []AccountInfo {
+	alist := make([]AccountInfo, len(a.c.Keys()))
+	for i, k := range a.c.Keys() {
+		v, _ := a.c.Get(k)
+		vp := v.(*AccountInfo)
+		alist[i] = *vp
+	}
+	return alist
+}
+
+func (a *AcctCache) GetAcctListOrg(orgId string) []AccountInfo {
+	var alist []AccountInfo
+	for _, k := range a.c.Keys() {
+		v, _ := a.c.Get(k)
+		vp := v.(*AccountInfo)
+		if vp.OrgId == orgId {
+			alist = append(alist, *vp)
+		}
+	}
+	return alist
+}
+
+func (a *AcctCache) GetAcctListRole(orgId, roleId string) []AccountInfo {
+	var alist []AccountInfo
+	for _, k := range a.c.Keys() {
+		v, _ := a.c.Get(k)
+		vp := v.(*AccountInfo)
+
+		if vp.RoleId == roleId && (vp.OrgId == orgId || OrgInfoMap.GetOrg(vp.OrgId).UltimateParent == orgId) {
+			alist = append(alist, *vp)
+		}
+	}
+	return alist
+}
+
+func (r *RoleCache) UpsertRole(orgId string, role string, voter bool, admin bool, access AccessType, active bool) {
+	defer r.mux.Unlock()
+	r.mux.Lock()
+	key := RoleKey{orgId, role}
+	r.c.Add(key, &RoleInfo{orgId, role, voter, admin, access, active})
+
+}
+
+func (r *RoleCache) GetRole(orgId string, roleId string) *RoleInfo {
+	defer r.mux.Unlock()
+	r.mux.Lock()
+	key := RoleKey{OrgId: orgId, RoleId: roleId}
+	if ent, ok := r.c.Get(key); ok {
+		return ent.(*RoleInfo)
+	}
+	return nil
+}
+
+func (r *RoleCache) GetRoleList() []RoleInfo {
+	rlist := make([]RoleInfo, len(r.c.Keys()))
+	for i, k := range r.c.Keys() {
+		v, _ := r.c.Get(k)
+		vp := v.(*RoleInfo)
+		rlist[i] = *vp
+	}
+	return rlist
+}
+
+// Returns the access type for an account. If not found returns
+// default access
+func GetAcctAccess(acctId common.Address) AccessType {
+	// check if the org status is fine to do the transction
+	a := AcctInfoMap.GetAccount(acctId)
+	if a != nil && a.Status == AcctActive {
+		// get the org details and ultimate org details. check org status
+		// if the org is not approved or pending suspension
+		o := OrgInfoMap.GetOrg(a.OrgId)
+		if o != nil && (o.Status == OrgApproved || o.Status == OrgPendingSuspension) {
+			u := OrgInfoMap.GetOrg(o.UltimateParent)
+			if u != nil && (u.Status == OrgApproved || u.Status == OrgPendingSuspension) {
+				if a.RoleId == networkAdminRole || a.RoleId == orgAdminRole {
+					return FullAccess
+				}
+				if r := RoleInfoMap.GetRole(a.OrgId, a.RoleId); r != nil && r.Active {
+					return r.Access
+				}
+				if r := RoleInfoMap.GetRole(o.UltimateParent, a.RoleId); r != nil && r.Active {
+					return r.Access
+				}
+			}
+		}
+	}
+	return DefaultAccess
+}
+
+func ValidateNodeForTxn(enodeId string, from common.Address) bool {
+	if enodeId == "" {
+		return true
+	}
+	ac := AcctInfoMap.GetAccount(from)
+	if ac == nil {
+		return true
+	}
+	ultimateParent := OrgInfoMap.GetOrg(ac.OrgId).UltimateParent
+	// scan through the node list and validate
+	for _, n := range NodeInfoMap.GetNodeList() {
+		if OrgInfoMap.GetOrg(n.OrgId).UltimateParent == ultimateParent && strings.Contains(n.Url, enodeId) {
```

Comment:
Defensively, I would use `enode.ParseV4(n.Url)` then compare the enodeId

Note: there's a type `enode.ID` to represent enodeId through out the code. Even though `string` doesn't create any issue, as far as consistency is concerned, `enode.ID` is preferred.

Reply:
Changed to enode.ID validation for this function alone. For rest of the places I have left it as string only as this is cache of the data in contract and in contract we are storing as string.